### DISCUSSION TITLE
it's possible to delete a api router after using include_router?

### DISCUSSION THREAD

**Discussion by gustavosett** [fastapi/fastapi#9855] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, APIRouter
from typing import Dict
import uvicorn

app = FastAPI()
routers: Dict[str, APIRouter] = {}

@app.get("/create_route/{route_name}")
async def create_route(route_name: str):
    router = APIRouter()

    @router.get(f"/{route_name}")
    async def new_route():
        return {"message": f"You have accessed {route_name} route"}

    routers[route_name] = router
    app.include_router(router)
    return {"message": f"Route {route_name} created"}

@app.get("/delete_route/{route_name}")
async def delete_route(route_name: str):
    del routers[route_name]
    # I can't delete the route from the app
    return {"message": f"Route {route_name} deleted"}


if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=8080, log_level="info")
```

### Description

I'm working on a system that requires dynamic routing. Currently, I am able to create a new route using **include_router**, but I am facing challenges when it comes to deleting a route **_without having to restart the server_**. It seems that there isn't a built-in 'delete route' functionality within FastAPI. I'm looking for a way to implement real-time route deletion. Can anyone guide me on how I might achieve this? Any help would be greatly appreciated.

### Operating System

Windows

### FastAPI Version

0.100.0

### Python Version

Python 3.11.3

**Comment by skarfie123** [comment]
I would not recommend this... But you could try to find and remove it from `app.router.routes`

**Reply by jamiekawabata** [reply]
Removing from `app.router.routes` causes the route to fail (as expected), but it does not remove it from the openapi spec, and subsequent adding of routes appears not to work.  It's effectively broken from that point on.

So it appears there's no way to remove routes.

**Comment by khaledadrani** [comment]
Hello, why do you need this? What is your use case?

**Comment by jamiekawabata** [comment]
I have a collection of files, each of which corresponds to an endpoint.  For the sake of argument let's say they are shell scripts or Lua programs or something, and each one has metadata describing the parameter structure and response structure.

The set of files can be constantly changing, even as the server is running, and the API needs to be updated to reflect that.

I found a solution within a comment on [this post](https://stackoverflow.com/questions/70783994/reload-routes-in-fastapi-during-runtime) on StackOverflow, namely "invalidate the cache and reload the new app specs"
`app.openapi_schema = None; app.setup()`

This worked, and my application is now working and updating the API dynamically.

I should mention, the API is not intended for a human to write client code against, since it is constantly shifting.  The idea is that an AI that understands OpenAPI definitions could scrape the API at the current moment, "decide" on an invocation, and make a well-formed request.

**Comment by krishnasism** [comment]
@jamiekawabata 
For regenerating the api spec dynamically, instead of using the default generated openapi spec, you can regenerate your openapi spec dynamically and serve it by overriding the "/docs" endpoint

```python
app = FastAPI(
    docs_url=None, redoc_url=None
)
```

```python
@app.get("/docs", include_in_schema=False)
async def serve_docs(request: Request):
   ....
```

There's some information on how to serve your custom HTML here on [stackoverflow](https://stackoverflow.com/questions/76406637/how-to-add-custom-html-content-to-fastapi-swagger-ui-docs)

**Reply by jamiekawabata** [reply]
Interesting.  Does that cover only the web page UI for testing?

It's not clear to me how customizing `/docs` would add or remove endpoints, i.e. what the server responds to.  In production, I am expecting the clients to pull `openapi.json` and then hit the endpoints based on that, and never use `/docs`.

Am I misunderstanding?

**Comment by YuriiMotov** [comment, marked as answer]
This can be achieved by combining [this](https://github.com/fastapi/fastapi/discussions/9855#discussioncomment-8779258) and this answers (delete routes and bypass the caching of openapi schema)
